Sources/CacheAdvance/CacheAdvance.swift

```diff
@@ -112,16 +112,18 @@ public final class CacheAdvance<T: Codable> {
 
         let cacheHasSpaceForNewMessageBeforeEndOfFile = writer.offsetInFile + bytesNeededToStoreMessage <= header.maximumBytes
         if header.overwritesOldMessages {
-            let truncateAtOffset: UInt64?
-            if cacheHasSpaceForNewMessageBeforeEndOfFile {
-                // We have room for this message. No need to truncate.
-                truncateAtOffset = nil
-            } else {
+            if !cacheHasSpaceForNewMessageBeforeEndOfFile {
                 // This message can't be written without exceeding our maximum file length.
                 // We'll need to start writing the file from the beginning of the file.
 
                 // Trim the file to the current writer position to remove soon-to-be-abandoned data from the file.
-                truncateAtOffset = writer.offsetInFile
+                try writer.truncate(at: writer.offsetInFile)
+
+                // Update the offsetInFileAtEndOfNewestMessage in our header and reader such that we will ignore the now-deleted data in the file.
+                // If the application crashes between writing this header and writing the next message data, we'll have only lost the messages we were about to delete.
+                // If `writer.offsetInFile == FileHeader.expectedEndOfHeaderInFile`, then crashing before we update the header would lead to this now-empty file being viewed as corrupted.
```

Comment:
I couldn't think of a way to eliminate this issue. If we truncate after writing the header, then we can still end up with file corruption because our `reader` doesn't actually know when the file _should_ stop before it reaches `maximumBytes`. So if we had:
```
[header][length|long-message7][length|message5][length|message6]garbage-remnant-of-message-3
                                                                ^ reader                      ^ maximumBytes
```

our `nextEncodedMessageSpan()` method would try to read the length of the the garbage remnant of the message three, and we'd end up trying to decode garbage.

Ideally our header would keep track of the last non-garbage byte in the file and we wouldn't need to truncate the file, but that's a much larger lift.

Reply:
> Ideally our header would keep track of the last non-garbage byte in the file and we wouldn't need to truncate the file, but that's a much larger lift.

This makes sense though I worry that there's a new chance of getting out of sync. Let's say we started with an empty file and write a message. We'd need to update both the offset of the end of the newest message, AND the offset of the last non-garbage byte in the file. We can't update both of those values atomically. In writing this comment I also realized (or re-realized) that none of our writes are guaranteed to be atomic, which is a scary thought.

Reply:
Yes the approach of keeping track of the last non-garbage byte in the file does lead to more chances that we get out of sync. However, we could:
1. Write the `header.offsetOfEndOfFile`. If we crash here we've just lost the data we were about to delete. We can accomplish this by making our reading code trust `offsetOfEndOfFile` more than it trusts `updateOffsetInFileAtEndOfNewestMessage` (i.e. `if offsetOfEndOfFile < updateOffsetInFileAtEndOfNewestMessage  { /* assume updateOffsetInFileAtEndOfNewestMessage is corrupt and trust offsetOfEndOfFile instead */ }`).
2. Write the `header.offsetInFileAtEndOfNewestMessage`. If we crash here we've just lost the data we were about to delete.

Basically I think adding this data to the header is workable, plugs a known hole in our existing logic, and would not lead to more potential file corruption. That said I don't think plugging this hole is worth the additional effort. Right now a crash during a write won't lead to data loss, which feels like the right bar to hit.